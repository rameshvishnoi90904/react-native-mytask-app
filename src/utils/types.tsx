export type task = {
  userId: number;
  id: number;
  title: string;
  body: string;
}

export type dispatchActionType = {
  type: string;
  data: task | null;
  searchInput: string;
}

export type NavigationScreenPropsType = {
  Home: undefined;
  TodoScreen: {item: task | null };
  LandingScreen: {};
};
