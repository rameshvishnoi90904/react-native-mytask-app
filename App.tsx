/**
 * Sample React Native App
 * https://github.com/facebook/react-native
 *
 * @format
 */

import React, { useEffect, useState } from 'react';
import {
  StyleSheet,
} from 'react-native';
import LaunchScreen from './src/screen/LaunchScreen';
import HomeScreen from './src/screen/HomeScreen';
import { NavigationScreenPropsType, task, dispatchActionType } from './src/utils/types';

import { NavigationContainer } from '@react-navigation/native';
import { createNativeStackNavigator } from '@react-navigation/native-stack';
import TodoDetailScreen from './src/screen/TodoDetailScreen';
import { AppContext } from './src/state/context';
import { fetchTodoList, deleteTodoItem, updateTodoItem, addNewTodoItem } from './src/api';

const Stack = createNativeStackNavigator<NavigationScreenPropsType>();

export const RootNavigator = () => {
  return (
    <Stack.Navigator>
      <Stack.Screen name="LandingScreen" component={LaunchScreen} />
      <Stack.Screen name="Home" component={HomeScreen} />
      <Stack.Screen name="TodoScreen" component={TodoDetailScreen} />
    </Stack.Navigator>
  )
}

function App(): JSX.Element {
  const initialTasks: Array<task> = [];
  const [tasks, setTasks] = useState(initialTasks);
  const [fullTaskList, setFullTaskList] = useState(initialTasks);
  useEffect(() => {
    fetchTodoList().then((value) => {
      setTasks(value);
      setFullTaskList(value);
    })
  },[])

  
  const dispatch = async (action: dispatchActionType) => {
    switch(action.type) {
      case 'delete': {
        const updatedTasks = tasks.filter((item) => {
          return item.id !== action?.data?.id;
        })
        if (action?.data?.id) {
          await deleteTodoItem(action.data.id)
        }
        setTasks(updatedTasks);
        setFullTaskList(updatedTasks)
      }
      break;

      case 'update': {
        const updatedTasks = [...tasks];
        if (action.data) {
          const toUpdateIndex = updatedTasks.findIndex((item) => {
            return item.id == action?.data?.id;
          });
  
          if (toUpdateIndex == -1) {
            const newTodo = await addNewTodoItem({...action.data, userId: 999});
            updatedTasks.splice(0, 0 , newTodo);
          } else {
            await updateTodoItem(action.data)
            updatedTasks[toUpdateIndex] = {...action.data}
          }
          setTasks(updatedTasks);
          setFullTaskList(updatedTasks)
        }
      }
      break;

      case 'search': {
        if (action.searchInput.length > 0) {
          const updatedTasks = fullTaskList.filter((item) => {
            const index = item.title.includes(action.searchInput);
            return index;
          })
          setTasks(updatedTasks);
        } else {
          setTasks(fullTaskList);
        }
      }
      break;

    }
  }
  
  return (
    <AppContext.Provider value={{tasks, dispatch}}>
        <NavigationContainer>
          {RootNavigator()}
      </NavigationContainer>
    </AppContext.Provider>

  );
}

const styles = StyleSheet.create({
  sectionContainer: {
    marginTop: 32,
    paddingHorizontal: 24,
  },
  sectionTitle: {
    fontSize: 24,
    fontWeight: '600',
  },
  sectionDescription: {
    marginTop: 8,
    fontSize: 18,
    fontWeight: '400',
  },
  highlight: {
    fontWeight: '700',
  },
});

export default App;
