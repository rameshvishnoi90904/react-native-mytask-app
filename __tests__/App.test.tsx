/**
 * @format
 */

import 'react-native';
import React from 'react';
import App from '../App';

// Note: import explicitly to use the types shiped with jest.
import {it, jest, expect} from '@jest/globals';

const mockTasks: Array<task> = [
  {id: 99, userId: 99, title: "sample title for testing", body: "sample body for testing"},
  {id: 98, userId: 99, title: "another sample user", body: "another sample body"}
];
jest.mock('../src/api/', () => ({
  fetchTodoList: () => new Promise((resolve, reject) => {
    resolve(mockTasks)
  }),
  deleteTodoItem: () => new Promise((resolve, reject) => {
    resolve(null)
  }),
  updateTodoItem: () => new Promise((resolve, reject) => {
    resolve(null)
  }),
  addNewTodoItem: (item: task) => new Promise((resolve, reject) => {
    resolve({...item, id: 999})
  })
}))

jest.spyOn(Alert, 'alert');


// Note: test renderer must be required after react-native.
import {create, act} from 'react-test-renderer';
import { fireEvent, render, screen, waitFor } from '@testing-library/react-native';
import { task } from '../src/utils/types';
import { Alert } from 'react-native';

it('renders correctly', () => {
    const tree = render(<App />).toJSON();
    expect(tree).toMatchSnapshot();
});

it('should render start button correctly', async () => {
  render(<App />);
  expect(screen.getByText('START')).toBeTruthy()
})


it('should navigate to home screen', async () => {
  render(<App />);
  fireEvent.press(screen.getByText('START'))

  expect(screen.getByText('+')).toBeTruthy()
})


it('dispatch delete event', async () => {
  render(<App />);
  act(() => {
    /* fire events that update state */
    fireEvent.press(screen.getByText('START'))
  });

  await waitFor(() => expect(screen.getByText("+")).toBeTruthy()) 

  fireEvent.press(screen.getByTestId(`todoItem-${mockTasks[0].id}`))

  act(() => {
    fireEvent.press(screen.getByText("Delete"))  
  })

  act(() => {
    Alert.alert.mock.calls[0][2][0].onPress()
  })
  await waitFor(() => expect(screen.getByText("+")).toBeTruthy()) 
  expect(screen.getByTestId('list-view').props.data).toStrictEqual([mockTasks[1]])
})


it('dispatch update event', async () => {
  render(<App />);
  act(() => {
    /* fire events that update state */
    fireEvent.press(screen.getByText('START'))
  });

  await waitFor(() => expect(screen.getByText("+")).toBeTruthy()) 

  fireEvent.press(screen.getByTestId(`todoItem-${mockTasks[0].id}`))

  act(() => {
    fireEvent.press(screen.getByText("Edit"))  
  })


  const updatedTitle = "Update task title triggered by testcase"
  fireEvent.changeText(screen.getByPlaceholderText("Enter your Title here"), updatedTitle)
  
  fireEvent.press(screen.getByText("Save"))

  await waitFor(() => expect(screen.getByText("+")).toBeTruthy()) 

  const updateTasks = [{...mockTasks[0], title: "Update task title triggered by testcase"}, mockTasks[1]]
  
  expect(screen.getByTestId('list-view').props.data).toStrictEqual(updateTasks)
})



it('dispatch update event for adding new task', async () => {
  render(<App />);
  act(() => {
    /* fire events that update state */
    fireEvent.press(screen.getByText('START'))
  });

  await waitFor(() => expect(screen.getByText("+")).toBeTruthy()) 

  fireEvent.press(screen.getByText("+"))

  // act(() => {
  //   fireEvent.press(screen.getByText("Edit"))  
  // })


  const newTaskTitle = "New task title triggered by testcase"
  fireEvent.changeText(screen.getByPlaceholderText("Enter your Title here"), newTaskTitle)
  

  const newTaskBody = "New Body title triggered by testcase"
  fireEvent.changeText(screen.getByPlaceholderText("Enter your Body here"), newTaskBody)
  

  fireEvent.press(screen.getByText("Save"))

  await waitFor(() => expect(screen.getByText("+")).toBeTruthy()) 

  const newlyAddedTask: task = {userId: 999, id: 999, title: newTaskTitle, body: newTaskBody};
  const updateTasks = [newlyAddedTask, ...mockTasks]
  
  expect(screen.getByTestId('list-view').props.data).toStrictEqual(updateTasks)
})




it('dispatch search event', async () => {
  render(<App />);
  act(() => {
    /* fire events that update state */
    fireEvent.press(screen.getByText('START'))
  });

  await waitFor(() => expect(screen.getByText("+")).toBeTruthy()) 

  const searchText = "user";
  act(() => {
    fireEvent.changeText(screen.getByPlaceholderText("Search here"), searchText)  
  })

  const matchingTasks = [mockTasks[1]]
  expect(screen.getByTestId('list-view').props.data).toStrictEqual(matchingTasks)
})
