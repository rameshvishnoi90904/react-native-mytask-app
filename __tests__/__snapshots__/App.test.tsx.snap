// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    {
      "flex": 1,
    }
  }
>
  <RNSScreenStack
    style={
      {
        "flex": 1,
      }
    }
  >
    <RNSScreen
      collapsable={false}
      gestureResponseDistance={
        {
          "bottom": -1,
          "end": -1,
          "start": -1,
          "top": -1,
        }
      }
      nativeBackButtonDismissalEnabled={false}
      onAppear={[Function]}
      onDisappear={[Function]}
      onDismissed={[Function]}
      onGestureCancel={[Function]}
      onHeaderBackButtonClicked={[Function]}
      onNativeDismissCancelled={[Function]}
      onTransitionProgress={[Function]}
      onWillDisappear={[Function]}
      replaceAnimation="push"
      sheetAllowedDetents="large"
      sheetCornerRadius={-1}
      sheetExpandsWhenScrolledToEdge={true}
      sheetGrabberVisible={false}
      sheetLargestUndimmedDetent="all"
      stackPresentation="push"
      style={
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
      swipeDirection="horizontal"
    >
      <View
        accessibilityElementsHidden={false}
        importantForAccessibility="auto"
        style={
          {
            "flex": 1,
            "flexDirection": "column-reverse",
          }
        }
      >
        <View
          collapsable={false}
          style={
            [
              {
                "flex": 1,
              },
              {
                "backgroundColor": "rgb(242, 242, 242)",
              },
              undefined,
            ]
          }
        >
          <RCTSafeAreaView
            style={
              {
                "alignItems": "center",
                "backgroundColor": "#EFEBE0",
                "flex": 1,
                "justifyContent": "center",
              }
            }
          >
            <View
              style={
                {
                  "flex": 1,
                }
              }
            />
            <View
              style={
                {
                  "alignItems": "center",
                  "flex": 1,
                  "justifyContent": "center",
                }
              }
            >
              <Text
                style={
                  {
                    "fontSize": 26,
                    "marginBottom": 16,
                  }
                }
              >
                Your Fav Todo App üìù
              </Text>
              <Text
                style={
                  {
                    "fontSize": 18,
                  }
                }
              >
                build by Ramesh
              </Text>
            </View>
            <View
              style={
                {
                  "flex": 1,
                  "justifyContent": "center",
                }
              }
            >
              <View
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": undefined,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  {
                    "alignItems": "center",
                    "backgroundColor": "#FB4570",
                    "borderRadius": 12,
                    "paddingVertical": 20,
                    "width": 200,
                  }
                }
                testID="start-button"
              >
                <Text
                  style={
                    {
                      "color": "#EFEBE0",
                      "fontSize": 22,
                      "fontWeight": "bold",
                    }
                  }
                >
                  START
                </Text>
              </View>
            </View>
          </RCTSafeAreaView>
        </View>
      </View>
      <RNSScreenStackHeaderConfig
        backButtonInCustomView={false}
        backTitleVisible={true}
        backgroundColor="rgb(255, 255, 255)"
        color="rgb(0, 122, 255)"
        direction="ltr"
        disableBackButtonMenu={false}
        hidden={false}
        hideBackButton={false}
        largeTitleHideShadow={false}
        title="LandingScreen"
        titleColor="rgb(28, 28, 30)"
        topInsetEnabled={false}
        translucent={false}
      />
    </RNSScreen>
  </RNSScreenStack>
</RNCSafeAreaProvider>
`;
